Glsl.meta.slang atomic #3587

jkwak-work opened the pull request:
PR for a review purpose.

ArielG-NV committed:
Implement all functions & syntax (with tests) for atomicOps and atomic_uint in opengl 4.6 spec

Implement all functions & syntax (with tests) for atomicOps and atomic_uint in opengl 4.6 spec, with extended type support for any shared standard opengl 4.6 ops inside extended type extensions (GL_EXT_shader_atomic_int64, GLSL_EXT_shader_atomic_float, GLSL_EXT_shader_atomic_float2)

Implementation details of worth{

atomic_uint in GLSL spec is not supported by Vulkan targets unless using a compiler extension (where it is then converted to a storage buffer). SPIRV->Vulkan does not support any form of AtomicCounter/AtomicStorage. As a result there is code to do what GLSL->Vulkan does, mangles atomic_uint into a storage buffer.

}

Testing details of worth{

Did not find a good way to test atomic_uint due to the storage buffer code conversion mangling names of objects, as a result we just test for compile

Float2 may be unsupported on most GPUs at the time of this PR, don't use it if it causes a hang/crash

}

ArielG-NV committed:
Merge branch 'shader-slang:master' into glsl.meta.slang_atomic_#3587

jkwak-work commented on source/slang/slang-ast-modifier.h:
```diff
@@ -733,6 +733,14 @@ class GLSLBindingAttribute : public Attribute
+{
+    SLANG_AST_CLASS(GLSLOffsetLayoutAttribute)
+
+    int offset;
```
It is the same thing but let's use int32_t here to keep the coding style same as the code around here.

jkwak-work commented on source/slang/slang-parser.cpp:
```diff
@@ -7797,12 +7860,25 @@ namespace Slang
                 modifier->loc = nameAndLoc.loc;
 
                 // Special handling for GLSLLayoutModifier
```
This comment will need to move down by three lines.

jkwak-work commented on source/slang/slang-parser.cpp:
```diff
@@ -7797,12 +7860,25 @@ namespace Slang
+                if (AdvanceIf(parser, TokenType::OpAssign))
                 {
-                    if (AdvanceIf(parser, TokenType::OpAssign))
+                    if (auto glslModifier = as<GLSLParsedLayoutModifier>(modifier))
```
It looks like you may be changing the behavior here.
Previously, if "modifier" is GLSLLayoutModifier and there is OpAssign, then it will "Advance".
But your new code will "Advance" for all of OpAssign even if "modifier" is not "GLSLLayoutModifier".

jkwak-work commented on source/slang/slang-diagnostic-defs.h:
```diff
@@ -243,6 +243,8 @@ DIAGNOSTIC(20101, Warning, unintendedEmptyStatement, "potentially unintended emp
 DIAGNOSTIC(30102, Error, declNotAllowed, "$0 is not allowed here.")
 
+DIAGNOSTIC(20001, Error, missingLayoutBindingModifier, "Expecting binding modifier with layout")
+
```
Is it possible to have a less cryptic error message?
If I encounter this error message in practice, I don't think I will know what I am supposed to do to fix it.

jkwak-work commented on source/slang/slang-emit-c-like.cpp:
```diff
@@ -2610,6 +2610,10 @@ void CLikeSourceEmitter::defaultEmitInstExpr(IRInst* inst, const EmitOpInfo& inO
         }
         break;
     }
+    case kIROp_RequireGLSLExtension:
+    {
+        break; //should already have set requirement; case covered for empty intrinsic block
+    }
```
This one is interesting one. I guess this fixes the strange behavior in your subgroup change?

ArielG-NV replied:
100%; I encountered the same problem now and since I understand what happens internally now I decided to fix the issue~ it was a simple problem of: kIROp_RequireGLSLExtension is not always possible to remove early (and emitter was not set to handle the op). Since this intrinsic was only used twice before this issue went unnoticed.

jkwak-work commented on source/slang/slang-emit-glsl.cpp:
```diff
@@ -2677,6 +2677,13 @@ void GLSLSourceEmitter::emitVarDecorationsImpl(IRInst* varDecl)
+    {
+        m_writer->emit(toSlice("layout(offset = "));
+        m_writer->emit(offsetDec->getOffset()->getValue());
+        m_writer->emit(toSlice(") "));
```
Would it make sense to have a new-line character at the end?

ArielG-NV replied:
we actually don't need this anymore since atomic_uint translation works properly, probably forgot to remove it after the initial change

jkwak-work commented on source/slang/slang-ir-inst-defs.h:
```diff
@@ -166,14 +166,15 @@ INST(Nop, nop, 0, 0)
     INST_RANGE(HLSLPatchType, HLSLInputPatchType, HLSLOutputPatchType)
 
-    INST(GLSLInputAttachmentType, GLSLInputAttachment, 0, HOISTABLE)
+    INST(GLSLInputAttachmentType, GLSLInputAttachment, 0, HOISTABLE)    
```
git system doesn't like trailing white characters at the end of lines.
They look same with and without them but they generate different hash values.

jkwak-work replied:
Can you follow the coding style around here.
It appears that each and every enum values are commented with their category.
Your new enum should either move to a place under a relevant comment or it should have its own new category.

jkwak-work commented on source/slang/slang-ir-inst-defs.h:
```diff
@@ -166,14 +166,15 @@ INST(Nop, nop, 0, 0)
         INST_RANGE(ByteAddressBufferTypeBase, HLSLByteAddressBufferType, HLSLRasterizerOrderedByteAddressBufferType)
         INST(RaytracingAccelerationStructureType, RaytracingAccelerationStructure, 0, HOISTABLE)
+        INST(GLSLAtomicUintType, GLSLAtomicUint, 0, HOISTABLE)
     INST_RANGE(UntypedBufferResourceType, HLSLByteAddressBufferType, RaytracingAccelerationStructureType)
```
This is an incorrect way to add a new enum here.
"INST_RANGE" is places right below their last enum value, it is "RaytracingAccelerationStructure" in this case.
You need to either move out "GLSLAtomicUintType" to elsewhere or include it to "UntypedBufferResourceType".

jkwak-work commented on source/slang/slang-ir-link.cpp:
```diff
@@ -1519,6 +1520,188 @@ static void diagnoseUnresolvedSymbols(TargetRequest* req, DiagnosticSink* sink,
+        // now we know the max offset and can construct a new storageBuffer with key of type maxOffset/4 (32bit uint)
+        auto elementType = builder.getArrayType(
+            builder.getUIntType(),
+            builder.getIntValue(builder.getUIntType(), uint32_t(maxOffset / 4)+uint32_t(1))
```
It will be better with
```
uint32_t(maxOffset/sizeof(uint32_t)+1)
```

jkwak-work commented on source/slang/slang-ir-link.cpp:
```diff
@@ -1519,6 +1520,188 @@ static void diagnoseUnresolvedSymbols(TargetRequest* req, DiagnosticSink* sink,
+
+                        // we fetch all global instances to pre-process
+                        auto layout = inst->findDecoration<IRLayoutDecoration>()->getLayout();
+                        auto layoutVal = as<IRVarOffsetAttr>(layout->getOperand(1))->getOffset();
```
I think it will be better to check if the return value from "as" is not null.
If the input source code by a mistake didn't have an operand to IRVarOffsetAttr, this would cause a null-pointer dereferencing, which can result in a silent crash or unexpected behaviors, isn't it?

jkwak-work commented on source/slang/slang-ir-link.cpp:
```diff
@@ -1519,6 +1520,188 @@ static void diagnoseUnresolvedSymbols(TargetRequest* req, DiagnosticSink* sink,
+
+void tryGLSLTypeReplacmentProcessing(IRSpecContext* context, TargetProgram* targetProgram, IRModule* irModule)
+{
+    if (targetProgram->getOptionSet().getBoolOption(CompilerOptionName::AllowGLSL))
```
Don't we have another way to take GLSL as input?
```
-lang glsl
```

ArielG-NV replied:
glsl.meta.slang only loads if AllowGLSL is true from what I understand

jkwak-work commented on source/slang/slang-ir-link.cpp:
```diff
@@ -1519,6 +1520,188 @@ static void diagnoseUnresolvedSymbols(TargetRequest* req, DiagnosticSink* sink,
+                }
+            }
+        }
+        ///
```
This seems like an unnecessary indentation.
There are papers that show better readability with indentation, but the effect diminishes when there are deep indentations.

jkwak-work commented on source/slang/slang-parser.cpp:
```diff
@@ -229,6 +229,24 @@ namespace Slang
+        }
+        int getNextBindingOffset(int binding) 
+        {
+            int& currentOffset = bindingToByteOffset.getOrAddValue(binding, 0);
```
I don't see why this uses a reference type.
It should be.,
```
int currentOffset = ...
```

jkwak-work commented on source/slang/slang-parser.cpp:
```diff
@@ -229,6 +229,24 @@ namespace Slang
+        {
+            bindingToByteOffset.set(binding, byteOffset);
+        }
+        int getNextBindingOffset(int binding) 
```
Wouldn't it make more sense to have const for this method?
```
int getNextBindingOffset(int binding) const
```

jkwak-work commented on source/slang/slang-parser.cpp:
```diff
@@ -4364,6 +4382,49 @@ namespace Slang
+        Modifiers* modifiers)
+    {
+        if (parser->options.allowGLSLInput) 
+        {
```
For this one, you can reduce the indentation by returning early.
```
if (!parser->options.allowGLSLInput)
    return;
```

jkwak-work replied:
But a better code will be to check the condition outside of the function.
```
if (parser->options.allowGLSLInput)
    AddSpecialModifierBasedOnType(...);
```
In general, it is not a good idea to evaluate a boolean condition inside of a function.
If you haven't, this will be a good book for your reference, "Clean Code - Robert C. Martin".

jkwak-work commented on source/slang/slang-type-layout.cpp:
```diff
@@ -3866,6 +3866,14 @@ static TypeLayoutResult _createTypeLayout(
             type,
             rules);
     }
+    else if (auto atomicType = as <GLSLAtomicUintType>(type))
```
Remove the space between "as" and "<".

jkwak-work commented on tests/glsl-intrinsic/atomic/atomicCounter.slang:
```diff
@@ -0,0 +1,142 @@
+//DISABLE_TEST:TEST(compute, vulkan):COMPARE_COMPUTE(filecheck-buffer=BUF):-vk -compute -entry computeMain -allow-glsl -emit-spirv-directly
+#version 430
+
+//TEST_INPUT:ubuffer(data=[0 0 0 0 0 0], stride=4):out
```
Is this line needed?
If yes, shouldn't it have a "name=XXX"?
Where is this going to be applied to?

ArielG-NV replied:
name is not necessary, if not used the binding will just be set on the next spot open, I will try a different method to get a buffer to bind properly with atomic_uint though

jkwak-work commented on tests/glsl-intrinsic/atomic/atomicCounter.slang:
```diff
@@ -0,0 +1,142 @@
+
+// currently tests do not handle atomic_uint->storage-buffer too well (mapping to binding location)
+//DISABLE_TEST:TEST(compute, vulkan):COMPARE_COMPUTE(filecheck-buffer=BUF):-vk -compute -entry computeMain -allow-glsl
+//DISABLE_TEST:TEST(compute, vulkan):COMPARE_COMPUTE(filecheck-buffer=BUF):-vk -compute -entry computeMain -allow-glsl -emit-spirv-directly
```
I think you should make these tests working before submit your change.
The input is GLSL and it should work for GLSL and SPIR-V functional testing.

jkwak-work commented on tests/glsl-intrinsic/atomic/atomicCounter.slang:
```diff
@@ -0,0 +1,142 @@
+// code is emitted correctly (this is a function call most importantly) 
+    return true
+        && atomicCounterExchange(param, 5) != 100
+// CHECK_GLSL-N: atomicCounterIncrement(one0._data_0[3]
```
Shouldn't this be one_0 with an underscore?

jkwak-work commented on tests/glsl-intrinsic/atomic/atomicCounter.slang:
```diff
@@ -0,0 +1,142 @@
+{
+    return true
+// CHECK_GLSL-DAG-N: atomicCounterExchange
+// CHECK_SPV-DAG-N: OpAtomicExchange
```
What is "-N" after DAG?
I don't see an explanation about it in the tutorial, https://llvm.org/docs/CommandGuide/FileCheck.html

ArielG-NV replied:
DAG-N means, if the following token shows up in any order at least once+previous counts of the DAG-N number of times then your token is found

needed since atomicCounterExchange will show up more than once and the function will likely be mangled some amount.

 LLVM FileCheck has a lot of undocumented/barley-documented components

jkwak-work replied:
The checking pattern is too lose.
We wouldn't be able to catch when there are problems.
Can you make more reliable tests to find any problems and regressions?

jkwak-work replied:
I will be careful to use rare syntax of FileCheck.
You also need to double check those syntax actually work with the version we use.
If I do something like
```
//CHECK_GLSL-DAG-MMM:
```
It will simply be considered as a comment and it wouldn't do anything.

ArielG-NV replied:
I noticed the following and removed all of the not properly working expressions

jkwak-work commented on tests/glsl-intrinsic/atomic/atomicStorageBuffer.slang:
```diff
@@ -0,0 +1,318 @@
+//TEST:SIMPLE(filecheck=CHECK_GLSL):  -allow-glsl -stage compute -entry computeMain -target glsl -DTARGET_GLSL
+//TEST:SIMPLE(filecheck=CHECK_SPV):  -allow-glsl -stage compute -entry computeMain -target spirv -emit-spirv-directly -DTARGET_SPIRV
+
+// currently tests do not handle atomic_uint well
```
This comment probably need to be removed.

jkwak-work commented on tests/glsl-intrinsic/atomic/atomicStorageBuffer.slang:
```diff
@@ -0,0 +1,318 @@
+#version 430
+
+
+//#define TEST_when_shader_atomic_float2_is_available
```
What is the story about this one? Why is it not available?
You may want to leave a descriptive explanation above the line.

ArielG-NV replied:
I will write an explanation; when testing if the code works with CI I will check if CI has atomic_float2 support and remove/keep this ifdef

jkwak-work commented on tests/glsl-intrinsic/atomic/atomicStorageBuffer.slang:
```diff
@@ -0,0 +1,318 @@
+    float data[1];
+} float32Buffer;
+
+//TEST_INPUT:ubuffer(data=[0 0], stride=8):out,name=float64Buffer
```
I think you need to use a double type expression for the data for all float variants.
```
data=[0.0]
```

jkwak-work commented on tests/glsl-intrinsic/atomic/atomicStorageBuffer.slang:
```diff
@@ -0,0 +1,318 @@
+        && atomicAdd(int32Buffer.data[0], 1) == 5
+        && int32Buffer.data[0] == 6
+
+        && (int32Buffer.data[0] = 5) == 5
```
This isn't really a part of "testing" but it is to re-initialize the value.
Maybe we can skip this repeated lines and chain the tests in a way that the previous values are used as a new initial value.

ArielG-NV replied:
we can't chain values because that would imply that every result must be the same (else inputs will be inconsistent when removing individual cases)

if every result is the same, we have no idea if the code is working, or doing nothing, since the tests are testing a scenario when both are possible

jkwak-work replied:
What I meant was more like this:
```
        && (int32Buffer.data[0] = 5) == 5
        && atomicAdd(int32Buffer.data[0], 1) == 5
        && int32Buffer.data[0] == 6
        && atomicMin(int32Buffer.data[0], 1) == 6
        && int32Buffer.data[0] == 1
        && atomicMax(int32Buffer.data[0], 5) == 5
        && int32Buffer.data[0] == 5
         ... and so on ....
```

ArielG-NV replied:
to feed last operations result into the next we need to assume the input->output will be the same always. 

the following will work predictably regardless of which block is commented out
```
        && (int32Buffer.data[0] = 5) == 5
        && atomicAnd(int32Buffer.data[0], 1) == 5
        && int32Buffer.data[0] == 1

        && (int32Buffer.data[0] = 5) == 5
        && atomicOr(int32Buffer.data[0], 2) == 5
        && int32Buffer.data[0] == 7

        && (int32Buffer.data[0] = 5) == 5
        && atomicXor(int32Buffer.data[0], 3) == 5
        && int32Buffer.data[0] == 6
```

but if we remove the setter we will have something like this which would work:
```
        && (int32Buffer.data[0] = 5) == 5
        && atomicAnd(int32Buffer.data[0], 1) == 5
        && int32Buffer.data[0] == 1

        && atomicOr(int32Buffer.data[0], 2) == 1
        && int32Buffer.data[0] == 2

        && atomicXor(int32Buffer.data[0], 3) == 2
        && int32Buffer.data[0] == 1
```

but if we comment out `Or` to test And and Xor we will get a test that is testing the wrong values:
```
        && (int32Buffer.data[0] = 5) == 5
        && atomicAnd(int32Buffer.data[0], 1) == 5
        && int32Buffer.data[0] == 1

        && atomicXor(int32Buffer.data[0], 3) == 2
        && int32Buffer.data[0] == 1
```

the setter makes this test stable for an inherently unstable operation

jkwak-work replied:
I agree that it will be good to be able to comment out a part of the chain.
But the initializing lines are too verbose and it is easy to overlook any mistakes like this,
```
        && (int32Buffer.data[0] = 5) == 5
        && atomicExchange(int32Buffer.data[0], 2) == 5
        && int32Buffer.data[0] == 2

        && (int32Buffer.data[0] = 6) == 5
        && atomicAnd(int32Buffer.data[0], 1) == 5
        && int32Buffer.data[0] == 1

        && (int32Buffer.data[0] = 5) == 5
        && atomicOr(int32Buffer.data[0], 2) == 5
        && uint32Buffer.data[0] == 7
```
Note that I replaced one of "5" with "6" in the middle, which is hard to spot if there was a bug like that; I planted one more bug.

I like to suggest to change the initialization as following,
```
        && i32_Init(5)
        && atomicExchange(int32Buffer.data[0], 2) == 5
        && i32_Expect(2)

        && i32_Init(5)
        && atomicAnd(int32Buffer.data[0], 1) == 5
        && i32_Expect(1)

        && i32_Init(5)
        && atomicOr(int32Buffer.data[0], 2) == 5
        && i32_Expect(7)
```

ArielG-NV replied:
👍

jkwak-work commented on tests/glsl-intrinsic/atomic/atomicStorageBuffer.slang:
```diff
@@ -0,0 +1,318 @@
+void computeMain()
+{
+    // testing has the following pattern in groups:
+    // set, op, test
```
I am not sure what this means.
Did you mean that "testAtomicInt32()", as an example, runs tests in a certain order?
I am not sure why it is important to be a comment.
And "set" sounds like a mutating functions, and "test" sounds like just testing the values but I cannot guess what "op" means.

ArielG-NV replied:
I will reword the comment

jkwak-work commented on tests/glsl-intrinsic/atomic/atomicStorageBuffer.slang:
```diff
@@ -0,0 +1,318 @@
+    uint64Buffer.data[0] = 2;
+    float16Buffer.data[0] = half(2.0f);
+    float32Buffer.data[0] = float(2.0f);
+    float64Buffer.data[0] = double(2.0);
```
I am not sure why these lines are needed.

ArielG-NV replied:
when checking the result buffer so there is only one value that is the value 1 (the results take all buffer outputs into account)

jkwak-work replied:
The result is stored in outputBuffer.data[0].
Why do we care about int32Buffer.data[0] and others?

jkwak-work reviewed (commented):
I have some comments.

jkwak-work reviewed (commented):
Left some comments.

jkwak-work commented:
Can we pass the atomic variables as a function argument?
I am curious how it is defined in HLSL and GLSL.
Probably Slang should do the same; and we should have a test case for it to prevent a regression.

I am also curious if it is allowed to have an atomic variable as a member variable in a struct.
And if the struct can be passed in to a function as an argument.
If yes, we should have a test cases for it.

jkwak-work commented:
I see compile error in the CI test, "Windows Build CI / build (Debug, Win32)"
```
D:\a\slang\slang\source\slang\slang-ir-link.cpp(1537,28): error C4244: 'initializing': conversion from 'Slang::IRIntegerValue' to 'int', possible loss of data [D:\a\slang\slang\build\visual-studio\slangc-bootstrap\slangc-bootstrap.vcxproj]
D:\a\slang\slang\source\slang\slang-ir-link.cpp(1538,36): error C4018: '<': signed/unsigned mismatch [D:\a\slang\slang\build\visual-studio\slangc-bootstrap\slangc-bootstrap.vcxproj]
D:\a\slang\slang\source\slang\slang-ir-link.cpp(1581,28): error C4244: 'initializing': conversion from 'Slang::IRIntegerValue' to 'int', possible loss of data [D:\a\slang\slang\build\visual-studio\slangc-bootstrap\slangc-bootstrap.vcxproj]
```

jkwak-work commented:
Also one of tests is failing in "Linux Build CI / build (release, gcc, x64)".
```
tests/glsl-intrinsic/atomic/atomicCounter.slang.1
```

ArielG-NV commented:
> I see compile error in the CI test, "Windows Build CI / build (Debug, Win32)"
> 
> ```
> D:\a\slang\slang\source\slang\slang-ir-link.cpp(1537,28): error C4244: 'initializing': conversion from 'Slang::IRIntegerValue' to 'int', possible loss of data [D:\a\slang\slang\build\visual-studio\slangc-bootstrap\slangc-bootstrap.vcxproj]
> D:\a\slang\slang\source\slang\slang-ir-link.cpp(1538,36): error C4018: '<': signed/unsigned mismatch [D:\a\slang\slang\build\visual-studio\slangc-bootstrap\slangc-bootstrap.vcxproj]
> D:\a\slang\slang\source\slang\slang-ir-link.cpp(1581,28): error C4244: 'initializing': conversion from 'Slang::IRIntegerValue' to 'int', possible loss of data [D:\a\slang\slang\build\visual-studio\slangc-bootstrap\slangc-bootstrap.vcxproj]
> ```

I will fix this after I correct atomic_uint implementation 👍 (fix all at once)

ArielG-NV commented:
> Can we pass the atomic variables as a function argument? I am curious how it is defined in HLSL and GLSL. Probably Slang should do the same; and we should have a test case for it to prevent a regression.

HLSL allows buffers as parameters (as an in parameter), not as a struct member. They have only atomic operations on storage buffer equivalents (byte address/RWBuffer), you cannot pass in a `inout int` to a RWBuffer, and then use that `int` as a atomic operation target

for GLSL you cannot pass buffers as parameters, *but* you can pass atomic_uint as a (in) parameter since it is an opaque pointer. atomic_uint also cannot be a member. What makes this unique is that atomic_uint is syntax candy for a storage buffer, but allows the additional functionality of using it as a parameter. From what I understand, if used, and if your compiler has the glsl_vk_relaxed extension, atomic_uint definition is converted to a storage buffer def, and all parameter->function uses of atomic_uint is converted to a uint reference of the element the atomic_uint would reference in the equivalent storage buffer implementation (either a forced early inline for all atomic_uint functions **or** it constructs a call graph and 'specializes' each function call for each atomic_uint in the program)

the use cases we need to test are as follows:
1. directly using in a function->call
2. as a function->function->call
3. as a function->function->function->...->call

I managed to get atomic_uint to work with the testing infrastructure, but I noticed nested parameters are an issue for my implementation since I rely heavily on ForceInline (likely need to find a way to restructure the code such that it works with unsafeForceInline)

ArielG-NV committed:
fix nested atomic_uint usage as function parameter; adjust syntax+style for slang repos corretness; fix errors from unused spirv ext; adjusted atomic subtraction operations to be correct for GLSL & SPIRV

jkwak-work commented on source/slang/slang-ir-link.cpp:
```diff
@@ -1534,7 +1532,7 @@ void convertAllAtomicUintToStructuredBufferUsages(
         for (auto& i : bindingToInstList.second)
         {
-            int currOffset = i->findDecoration<IRGLSLOffsetDecoration>()->getOffset()->getValue();
+            uint32_t currOffset = uint32_t(i->findDecoration<IRGLSLOffsetDecoration>()->getOffset()->getValue());
```
I think maxOffset and currOffset should be 64bit.

ArielG-NV replied:
👍 32bit is a mistake, vkdevice limits allow more than 4GB worth of data

jkwak-work commented on source/slang/slang-parser.cpp:
```diff
@@ -235,9 +235,9 @@ namespace Slang
+        const int getNextBindingOffset(int binding) 
         {
-            int& currentOffset = bindingToByteOffset.getOrAddValue(binding, 0);
+            const int currentOffset = bindingToByteOffset.getOrAddValue(binding, 0);
```
Because of the name of the function, I got an impression that this function doesn't mutate "this".
But the function "getOrAddValue()" will mutate when it add.
So you cannot apply "const" to the function.
What I suggest was
```
int getNextBindingOffset(int binding) const
```
But you cannot do it.
Adding "const" to the return type has no meaning when the type is a primitive type; please remove it.

ArielG-NV replied:
getNextBindingOffset mutates a dictionary, but the value returned is const, it should never be changed once returned, only used or discarded

jkwak-work replied:
It seems that you are not vary familiar with how "const" modifier works.
I recommend you to checkout a fantastic book called, "Effective C++ - Scott Meyers" if you haven't.

Note that you cannot force the calling site to treat the return value as const as you intended.
There will be no error nor warning when you do something like,
```
const int getNextBindingOffset(int binding) { ...}
int a = getNextBindingOffset(...); // no warning no error.
```

jkwak-work commented on tests/glsl-intrinsic/atomic/atomicCounter.slang:
```diff
@@ -48,87 +43,79 @@ bool testAtomicUint()
-// CHECK_GLSL-DAG-N: atomicCounterIncrement(one0._data_0[3]
-// CHECK_SPV-DAG-N: OpAtomicIAdd
+// CHECK_GLSL-N: atomicCounterIncrement(one_0._data_0[3]
+// CHECK_SPV-N: OpAtomicIAdd
```
One of my concerns about this is that OpAtomicIAdd may find a matching line generated from the prvious functions.
I think you can isolate the searching range by checking the function name first, something like this:
```
// CHECK_SPV-LABEL: counterAsParam
// CHECK_SPV: OpAtomicIAdd

// CHECK_SPV-LABEL: testAtomicUint
// CHECK_SPV: OpAtomicIAdd
```
You can use LABEL and drop the "-N" for OpAtomicIAdd.

ArielG-NV replied:
I will change the syntax, your concern is 100% correct, the tests sometimes pass when they should fail. I will remedy this (using label and changing the tests), I think having good tests for atomic_uint will be quite helpful since its a big compiler magic trick

jkwak-work commented on tests/glsl-intrinsic/atomic/atomicStorageBuffer.slang:
```diff
@@ -1,57 +1,57 @@
 
-
-//#define TEST_when_shader_atomic_float2_is_available
+// float2 is currently a very new extension; most hardware lacks 
+// this extension and will fail the test if attempting to use atomic_float2 
+// operations
```
Which hardware would it work with?
Isn't Ampere (RTX 30XX) new enough?

ArielG-NV replied:
https://vulkan.gpuinfo.org/displayreport.php?id=28714#device
likely just not added (2024-march 4th)

jkwak-work replied:
The spec might be newer than hardware.
But it doesn't mean the older hardware wouldn't support.

I am not sure what the link you posted.

jkwak-work replied:
I see that the link shows which extensions are supported for 4090.
And I don't see "VK_EXT_shader_atomic_float" in it.

ArielG-NV replied:
![image](https://github.com/jkwak-work/slang/assets/159081215/5c58b16c-edca-4828-8929-235dce9d9cb5)

ArielG-NV replied:
(missing atomic_float2)

jkwak-work reviewed (commented):
Minor comments and questions

ArielG-NV committed:
use proper offset type (int64, since max size is greater than 4gb for storage buffers in vk)

change tests to test properly for op codes

jkwak-work commented on source/slang/glsl.meta.slang:
```diff
@@ -2825,17 +2825,13 @@ public uint rayQueryGetIntersectionTypeEXT(rayQueryEXT q, bool committed)
-void typeRequireChecks_shader_subgroup_GLSL() {
-    // the following is a seperate function call, since else the `__requireGLSLExtension` and associated __intrinsic_asm is ignored if the calling function also calls an __intrinsic_asm
+void shader_subgroup_preamble() 
+{
```
This should be in a separate change not a part of "atomic" change.

jkwak-work replied:
Are you going to keep this change here?

jkwak-work commented on source/slang/glsl.meta.slang:
```diff
@@ -3088,7 +3051,6 @@ __spirv_version(1.3) [require(spirv)]
             OpCapability Shader;
             OpControlBarrier Subgroup Subgroup AcquireRelease|SubgroupMemory|ImageMemory|UniformMemory
         };
-
```
Removing or adding new lines is just a cosmetic change.
Try not to do it as much as you can resist in a sense that your change should be minimal for the clarity of the purpose.

This type of cosmetic harmless change can cause conflicts with other harmless cosmetic changes of unrelated changes done by other people. It can be a bigger problem in bigger projects.

jkwak-work replied:
I suggested not to remove this line and are you going to keep it as is?

jkwak-work commented on source/slang/glsl.meta.slang:
```diff
@@ -4615,4 +4574,607 @@ __spirv_version(1.3) [require(spirv)]
+__intrinsic_type($(kIROp_GLSLAtomicUintType))
+public struct atomic_uint{};
+
+// teir of float refers to atomic extension support of float1 or float2.
```
Did you mean tier?

ArielG-NV replied:
I did mean tier (...)

jkwak-work commented on source/slang/glsl.meta.slang:
```diff
@@ -4615,4 +4574,607 @@ __spirv_version(1.3) [require(spirv)]
+    {
+        if (__type_equals<T, float>())
+        {
+            spirv_asm
```
Don't you need to return?
```
return spirv_asm
```

ArielG-NV replied:
the return would be void, I just need SPIR-V ASM to be passed to emit-spirv to have them evaluated and pushed to extension+capability scope of SPIR-V

jkwak-work commented on source/slang/glsl.meta.slang:
```diff
@@ -4615,4 +4574,607 @@ __spirv_version(1.3) [require(spirv)]
+    const char* classType;
+    const char *subclassType;
+    const char *suffix;
+} atomics[] =
```
From the readability perspective, it will be better to make this two statements;
```
struct XXX {
...
};
static const XXX atomics[] = { ... };
```

jkwak-work commented on source/slang/glsl.meta.slang:
```diff
@@ -4615,4 +4574,607 @@ __spirv_version(1.3) [require(spirv)]
+for (int i = 0; i < 7; i++) 
+{
+    TypeClassSubclass item = atomics[i];
+}}}}
```
You can avoid using the hardcoded value 7.
I think we have ARRAY_SIZE(atomics) or something similar.
I think you may even be able to do
```
for ( auto atomic : atomics)
```
Give a try.

jkwak-work replied:
Are you still going to use the magic numbers?

ArielG-NV committed:
handle slang parsing errors more elegantly with assign op inside layout (needed since future commit will be using same code; will be less hard to deal with code conflicts)

jkwak-work commented on source/slang/glsl.meta.slang:
```diff
@@ -4615,4 +4574,607 @@ __spirv_version(1.3) [require(spirv)]
+// converted into one as such
+__magic_type(GLSLAtomicUintType)
+__intrinsic_type($(kIROp_GLSLAtomicUintType))
+public struct atomic_uint{};
```
This looks little confusing.
Let's make it a multi-line,
```
public struct atomic_uint
{
};
```

jkwak-work replied:
Are you going to keep it as is?

jkwak-work commented on source/slang/glsl.meta.slang:
```diff
@@ -4615,4 +4574,607 @@ __spirv_version(1.3) [require(spirv)]
+if(0
+    || i == 3
+    || i == 4
+    || i == 5
```
This magic numbers can cause a problem later when other programmer make changes code around here without realizing what these numbers are for.
It will be better to add a new member variable to TypeClassSubclass and use the descriptive variable name here.
```
struct TypeClassSubclass
{
....
    bool supportExtraOrWhatever
...
};
...
if (!atomic.supportExtraOrWhatever) continue;
```

jkwak-work replied:
Are you going keep the magic numbers here too?

jkwak-work commented on source/slang/slang-ir-link.cpp:
```diff
@@ -1519,6 +1520,146 @@ static void diagnoseUnresolvedSymbols(TargetRequest* req, DiagnosticSink* sink,
+
+        // our "_data" struct member
+        auto elementBufferKey = builder.createStructKey();
+        builder.addNameHintDecoration(elementBufferKey, UnownedStringSlice("_data"));
```
Please either remove the comment or explain more of "why" we need "_data".

jkwak-work replied:
Please remove this comment.

jkwak-work commented on source/slang/slang-ir-link.cpp:
```diff
@@ -1519,6 +1520,146 @@ static void diagnoseUnresolvedSymbols(TargetRequest* req, DiagnosticSink* sink,
+            for (auto use = i->firstUse; use; use = next)
+            {
+                next = use->nextUse;
+                auto user = use->user;
```
I am failing to see why "use->nextUse" has to be stored in "next" prior to the processing.
The body of for-loop uses "use->user" and there is no way for "use->nextUse" can be altered or even accessed in the middle of the for-loop.

jkwak-work replied:
Please explain why you are making a temporary variable unnecessarily.

ArielG-NV replied:
since we are using `replaceUsesWith`, emitting types, and setting types/operands, we are adding IRUses to an eventual parent atomic_uint->storageBuffer and changing the list head. If we do not store the old node we will be changing our IRUse list and be losing the `->next` node. As a result our code will not mangle all atomic_uint references and will stop after the first.

ArielG-NV replied:
this can be seen with the code emitted (or vulkan layers complaining that we have some 'ptr' to an int

I think there was a second reason I did this, but it has been a week and I don't remember; I do recall it being crash related though (node1->node2->node3 became node3->node1->node2, and then node2->node3->node1, causing an infinite loop)

jkwak-work commented on source/slang/slang-ir-link.cpp:
```diff
@@ -1519,6 +1520,146 @@ static void diagnoseUnresolvedSymbols(TargetRequest* req, DiagnosticSink* sink,
+    }
+}
+
+void tryGLSLTypeReplacmentProcessing(IRSpecContext* context, TargetProgram* targetProgram, IRModule* irModule)
```
My understanding is that "tryXXX" functions return bool type so that if the attempt was successful it will return true; otherwise false.
I think you can take off "tryGLSL"

jkwak-work replied:
Please rename the function.

jkwak-work commented on source/slang/slang-ir-link.cpp:
```diff
@@ -1519,6 +1520,146 @@ static void diagnoseUnresolvedSymbols(TargetRequest* req, DiagnosticSink* sink,
+    // this is an issolated boolean operation since this operation can be moved to different
+    // parts of Slang and may likley need to be moved at some point
+    if (targetProgram->getOptionSet().getBoolOption(CompilerOptionName::AllowGLSL))
+    {
```
You can reduce the indentation of the whole body of the function with a following code,
```
if (!targetProgram->...AllowGLSL)) return;
```

jkwak-work commented on source/slang/slang-ir-link.cpp:
```diff
@@ -1519,6 +1520,146 @@ static void diagnoseUnresolvedSymbols(TargetRequest* req, DiagnosticSink* sink,
+void tryGLSLTypeReplacmentProcessing(IRSpecContext* context, TargetProgram* targetProgram, IRModule* irModule)
+{
+    // this is an issolated boolean operation since this operation can be moved to different
+    // parts of Slang and may likley need to be moved at some point
```
This should be either removed or be "TODO: relocate to other place for XXX reason"

jkwak-work replied:
Please remove this comment.

ArielG-NV replied:
I will remove the comment; the idea was to explain why I have an isolated boolean since I understand it is the wrong way to go about writing code (at least in most cases, as per earlier review)

jkwak-work commented on source/slang/slang-ir-link.cpp:
```diff
@@ -1519,6 +1520,146 @@ static void diagnoseUnresolvedSymbols(TargetRequest* req, DiagnosticSink* sink,
+    // parts of Slang and may likley need to be moved at some point
+    if (targetProgram->getOptionSet().getBoolOption(CompilerOptionName::AllowGLSL))
+    {
+        /// do 1 pass and get all data needed to process GLSL inst's that need attention
```
Please remove this comment.
"get all data needed attention" is too obvious and it is not useful.

jkwak-work replied:
I don't think we use three slashes for commenting.
Please follow the coding style of what has been already used.
As far as coding style goes, it should be "as if one person wrote the whole code".

ArielG-NV replied:
should we at some point have some pr that fixes the sections of code that do not follow Slang coding style? Lots of the code is inconsistent, and it is causing a huge problem for me when trying to figure out how to go about X and Y when Z does something different for the very specific case that is Z

jkwak-work commented on source/slang/slang-ir-link.cpp:
```diff
@@ -1519,6 +1520,146 @@ static void diagnoseUnresolvedSymbols(TargetRequest* req, DiagnosticSink* sink,
+                {
+                    // Atomic_uint are supported by GLSL through converting to a different type (compiler ext)
+                    // not supported by SPIR-V->VK, means we must convert the type ourselves, else SPIR-V + 
+                    // HLSL is going to be impossible to emit as a target
```
I am having trouble understanding what this comment means.
What does "SPIR-V->VK" and "SPIR-V + HLSL" mean?
As far as I know, we don't translate from Spir-V to Vulkan since there is no Vulkan specific shader language. I guess that you were using "->" for a meaning other than "translate/convert"?

This file is at IR level and I don't think it is appropriate to talk about the feasibility of emit targets.

jkwak-work replied:
Please explain to my question

ArielG-NV replied:
I actually do mean '->' as in translate/convert/compile. SPIR-V->VK is an exact meaning. SPIR-V has no issues with AtomicCounter and AtomicStorage, it supports both. The issue is Vulkan spec dictating that atomic_uint and equivalents are not allowed in Vulkan, meaning when any driver is compiling SPIR-V->driver the code will fail to compile (SPIR-V compilers will throw an error if you let it to and tell it VK is the target) if using an atomic_uint equivlent in SPIR-V (AtomicCounter/AtomicStorage) 

This IR should bring up emit targets from what I understand since it will directly impact the emitted code (emit is supposed to only emit, not mangle types or do whole code passes), we need this code if we want to have the ability to emit [HLSL|SPIRV|GLSL]->VK/D3D12

jkwak-work commented on source/slang/slang-ir-link.cpp:
```diff
@@ -1519,6 +1520,146 @@ static void diagnoseUnresolvedSymbols(TargetRequest* req, DiagnosticSink* sink,
+                    // not supported by SPIR-V->VK, means we must convert the type ourselves, else SPIR-V + 
+                    // HLSL is going to be impossible to emit as a target
+
+                    // we fetch all global instances to pre-process
```
Please remove this.

jkwak-work replied:
Please remove this .

ArielG-NV committed:
let read token handle errors and keep the removed token peek (caused incorrect code parsing)

jkwak-work commented on source/slang/slang-ir-link.cpp:
```diff
@@ -1519,6 +1520,146 @@ static void diagnoseUnresolvedSymbols(TargetRequest* req, DiagnosticSink* sink,
+                };
+            }
+        }
+        ///
```
Please remove unnecessary "///"

jkwak-work replied:
Please remove this comment.

ArielG-NV replied:
I suppose I will explain this topic just for future reference; GLSL has lots of hacks and random things which are removed if GLSL->VK is the target. GLSL also has lots of odd styling which makes little sense when translating GLSL->[SPIRV|CUDA|HLSL], as a result there will need to be some styling at some point to make this section less cluttered (since it likely will be used more)

jkwak-work commented on tests/glsl-intrinsic/atomic/atomicCounter.slang:
```diff
@@ -0,0 +1,130 @@
+    return true
+    
+        && atomicCounterExchange(one, 5) != 100
+// CHECK_GLSL-DAG: atomicAdd(
+// CHECK_SPV-DAG: OpAtomicIIncrement
```
I think you can start by finding the function name first so that your pattern search doesn't find the matching from the previous function.
```
CHECK_GLSL: testAtomicUint
CHECK_SPV: testAtomicUint
```
Or you can use XXX-LABEL

jkwak-work replied:
Please add the suggested lines.

ArielG-NV replied:
this is no longer an issue since each function shows up only a limited number of times, if I check for add twice, and I find add twice, we know the code is working?

ArielG-NV replied:
I will add a label check to ensure order though

jkwak-work commented on source/slang/slang-ir-link.cpp:
```diff
@@ -1519,6 +1520,146 @@ static void diagnoseUnresolvedSymbols(TargetRequest* req, DiagnosticSink* sink,
+
+    for (auto& bindingToInstList : bindingToInstMapUnsorted)
+    {
+        //I don't need a sorted list, just the max so I can make an array of uint[(maxOffset/4)+1]
```
Please remove this comment.
It simply repeats what the code is doing.
I like to remind you that our coding style states the following,
```
Try to write comments that explain the "why" of your code more than the "what".
```

jkwak-work replied:
Please remove this comment.

ArielG-NV replied:
I will remove the comment; my point with the comment was to explain a why, specifically why we are not sorting the array -- we just need the max and order is not relevant (but I understand why that may be not needed)

jkwak-work commented on source/slang/slang-ir-link.cpp:
```diff
@@ -1519,6 +1520,146 @@ static void diagnoseUnresolvedSymbols(TargetRequest* req, DiagnosticSink* sink,
+        auto instToSwitch = *bindingToInstList.second.begin();
+        builder.setInsertBefore(instToSwitch);
+        
+        // now we know the max offset and can construct a new storageBuffer with key of type maxOffset/4 (32bit uint per atomic_uint)
```
Please remove this comment

jkwak-work replied:
Please remove this comment.

jkwak-work commented on source/slang/slang-ir-link.cpp:
```diff
@@ -1519,6 +1520,146 @@ static void diagnoseUnresolvedSymbols(TargetRequest* req, DiagnosticSink* sink,
+            builder.getIntValue(builder.getUIntType(), (maxOffset / sizeof(uint32_t))+1)
+        );
+
+        // now we make our struct which goes within with some name (does not matter), 
```
Please remove this comment

jkwak-work replied:
Please remove this comment.

jkwak-work commented on source/slang/slang-ir.cpp:
```diff
@@ -4495,6 +4495,18 @@ namespace Slang
 
+    IRGLSLShaderStorageBufferType* IRBuilder::createGLSLShaderStorableBufferType(UInt operandCount, IRInst* const* operands)
+    {
+        IRGLSLShaderStorageBufferType* ssboType = createInst<IRGLSLShaderStorageBufferType>(
+            this,
+            kIROp_GLSLShaderStorageBufferType,
+            getTypeKind(),
+            operandCount,
+            operands);
```
Shouldn't this be like this?
```
        IRGLSLShaderStorageBufferType* ssboType = createInst<IRGLSLShaderStorageBufferType>(
            this,
            kIROp_GLSLShaderStorageBufferType,
            getTypeKind(),
            operands[0],
            operands[1]);
```
You should try to avoid using a variable-length array as a function parameter if possible.

jkwak-work replied:
Please answer to my question

ArielG-NV replied:
We should not use a fixed array. Operands in this scenario represent members and modifiers in/of a storage buffer. Storage buffers in GLSL allow multiple modifiers, if we do not allow a variable length, this code makes little sense

to explain the styling: `IRInterfaceType* IRBuilder::createInterfaceType(UInt operandCount, IRInst* const* operands)` is the closest to a struct we have in the IR create side of things. This code works the exact same as an interface construction, which is likely how the code should be designed since an interface is a parallel (different functionality but similar in syntax) to a struct

jkwak-work commented on source/slang/slang-ir-link.cpp:
```diff
@@ -1519,6 +1520,146 @@ static void diagnoseUnresolvedSymbols(TargetRequest* req, DiagnosticSink* sink,
+        auto elementBufferType = elementType;
+        auto _dataField = builder.createStructField(structType, elementBufferKey, elementBufferType);
+        
+        // make our valid storage buffer type
```
Please remove this comment.

jkwak-work replied:
Please remove this comment.

jkwak-work commented on source/slang/slang-ir-link.cpp:
```diff
@@ -1519,6 +1520,146 @@ static void diagnoseUnresolvedSymbols(TargetRequest* req, DiagnosticSink* sink,
+        
+        // now we need to make all call's to:
+        // 1. convert all atomic_uint function type parameters to a inlined call of atomic_uint->inout int ref of 
+        // field->element-addr based on location(offset) from the atomic_uint->storage_buffer logic above
```
I see "1." but I don't see "2.".
And this needs to be either removed or simplified to be helpful.

jkwak-work replied:
Please remove this comment.

ArielG-NV replied:
I will simplify the comment as per the first review, 

these comments are here since the code here was trying to follow how slang-ir-append-consume-struct was commented since it similarly is pulling a similar style of 'magic trick' to manipulate var A into var B and var C  (it has some 'whats' rather than 'whys')

ArielG-NV replied:
// Atomic_uint definitions needs to become a storage buffer to follow GL_EXT_vulkan_glsl_relaxed
    // and to allow translation of atomic_uint into SPIRV
    
    // All references to a atomic_uint need to be an element ref. to emulate storage buffer usage
   
    // All function calls must be inlined since storage buffers cannot pass as parameters to atomic methods

will be the new comments for this function and will be placed accordingly. If these are not placed I am without doubt sure there will be confusion on why any of this code exists, it heavily mangles the inputted code.

jkwak-work commented on source/slang/slang-parser.cpp:
```diff
@@ -4364,6 +4382,46 @@ namespace Slang
+            if (auto varDeclBase = as<VarDeclBase>(decl))
+            {
+                if (auto declRefExpr = as<DeclRefExpr>(varDeclBase->type.exp))
+                {
```
I don't see why your comment is needed here.
Calling an expensive function later is a common practice in programming.

And I like to suggest the following code for less indentation.
```
auto varDeclBase = as<VarDeclBase>(decl));
if (!varDeclBase) return;

auto declRefExpr = as<DeclRefExpr>(varDeclBase->type.exp);
if (!declRefExpr) return;

auto name = declRefExpr->name;
if (!name) return;
if (!name->text.equals("atomic_uint")) return;

auto bindingMod = modifiers->findModifier<GLSLBindingAttribute>();
if (!bindingMod) return;
```

jkwak-work commented on source/slang/slang-parser.cpp:
```diff
@@ -4364,6 +4382,46 @@ namespace Slang
+                                
+                                Modifiers newModifier;
+                                newModifier.first = modifier;
+                                _addModifiers(decl, newModifier);
```
Do we have a test coverage for this code and the examples mentioned in the comment?
If not, can you add some?

jkwak-work commented on source/slang/glsl.meta.slang:
```diff
@@ -4615,4 +4574,607 @@ __spirv_version(1.3) [require(spirv)]
+__glsl_version(430) [require(glsl)]
+[ForceInline] public uint atomicCounterDecrement_GLSL_wrapper(atomic_uint c)
+{
+    __target_switch
+    {
+    case glsl:
+    {
+        __intrinsic_asm "atomicExchange($0,$0-1)";
+    }
```
Do you need __target_switch when it requires only glsl?

And this function is used only once.
It will be better to place near where it is called from.

And the function doesn't "wrap" anything.
I would say "_helper".

ArielG-NV replied:
I need the function since __intrinsic_asm returns (and we must return using atomicCounter to get the decrement logic correct)

I will rename the '_wrapper' to '_helper'

jkwak-work commented on source/slang/glsl.meta.slang:
```diff
@@ -4615,4 +4574,607 @@ __spirv_version(1.3) [require(spirv)]
+//// GLSL atomic
+
+// this is actaully just a fancy storage buffer, and will be 
+// converted into one as such
```
You should try to stay in "impartiality tone", which is objective and formal.
I would say more like as following,
```
// atomic_uint is internally implemented with Shader Storage Object Buffer.
```

jkwak-work commented on source/slang/slang-parser.cpp:
```diff
@@ -229,6 +229,24 @@ namespace Slang
+            const int64_t currentOffset = bindingToByteOffset.getOrAddValue(binding, 0);
+            bindingToByteOffset.set(
+                binding,
+                currentOffset+4
```
Can this magic number be replaced with "sizeof(uint)"?

jkwak-work reviewed (commented):
I have some comments.
Mostly for the coding style.

ArielG-NV committed:
fix bug with offset not properly managed from zero'ed position

add test for multiple offset atomic counters and verification

adjust tests for syntax; work arounding a bug found (will make issue/and-or fix)

ArielG-NV committed:
fix mistake in glsl.meta (array size 6 when meant to be 7 -- warning->error)

jkwak-work commented on source/slang/slang-ast-type.h:
```diff
@@ -252,7 +252,6 @@ class RaytracingAccelerationStructureType : public UntypedBufferResourceType
     SLANG_AST_CLASS(RaytracingAccelerationStructureType)
 };
 
-
```
Please don't remove lines unnecessarily.

jkwak-work reviewed (commented):
Most of my suggestion are not answered.

ArielG-NV committed:
adjustment of syntax
